### PR TITLE
feat: allow SQL Lambda to be hotswapped in sandbox

### PR DESCRIPTION
**Pull request by bobbyu99**
#### Description of changes
Added a deployment identifier to differentiate between a sandbox and branch deployment. If it is a sandbox deployment, `SQLLayerVersionCustomResource` and `SQLSNSTopicARNCustomResource` will use a physical ID in the format of `resourceName-YYYY-MM-DD` to avoid multiple deployments in the same day. 

#### Description of how you validated changes
Unit test added.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] Any CDK or CloudFormation parameter changes are called out explicitly

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts [hunk 1]
```diff
@@ -196,14 +196,28 @@ export const createRdsLambda = (
...
   const key = Fn.join('', [SQLLayerVersionManifestKeyPrefix, Fn.ref('AWS::Region')]);
 
   const manifestArn = `arn:aws:s3:::${SQLLayerManifestBucket}/${key}`;
 
   const resourceName = resourceNames.sqlLayerVersionResolverCustomResource;
+
+  // if deployment type is sandbox, use id in the format: resourceName-YYYY-MM-DD to avoid multiple deployments in the same day
+  // if deployment type is branch, use id in the original format
+  let physicalIdValue: string;
+  if (context?.deploymentIdentifier?.deploymentType === 'sandbox') {
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:215]: nit: consider a [TS type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates) to determine whether you have a sandbox type or not. That has the advantage of centralizing type introspection into a single function rather than scattering it throughout the project. It also allows TypeScript to infer the structure of the rest of the object even if the type structure gets hairy. See examples in `packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts` ([example](https://github.com/aws-amplify/amplify-category-api/blob/main/packages/amplify-graphql-transformer-core/src/utils/model-datasource-strategy-utils.ts#L76-L87))

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts [hunk 2]
```diff
@@ -196,14 +196,28 @@ export const createRdsLambda = (
...
+    physicalIdValue = `${resourceName}-${new Date().toISOString().substring(0, 10)}`;
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:216]: Why is "per day" the right granularity for controlling deployments in a sandbox? Why would we try to compromise on this at all? By introducing an arbitrary cutoff, we're going to still require deployments for some customers whose sandbox deployments happen to span a day boundary, while missing layer version updates for customers who do multiple deployments that don't happen to span a day boundary. Since sandboxes aren't intended for production, why can't we just use a strictly deterministic resource name, based e.g. on the sandbox name?

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts [hunk 3]
```diff
@@ -231,14 +245,28 @@ export const createLayerVersionCustomResource = (scope: Construct, resourceNames
...
+): AwsCustomResource => {
   const { SQLLayerManifestBucket, SQLLayerManifestBucketRegion, SQLSNSTopicARNManifestKeyPrefix } = ResourceConstants.RESOURCES;
 
   const key = Fn.join('', [SQLSNSTopicARNManifestKeyPrefix, Fn.ref('AWS::Region')]);
 
   const manifestArn = `arn:aws:s3:::${SQLLayerManifestBucket}/${key}`;
 
   const resourceName = resourceNames.sqlSNSTopicARNResolverCustomResource;
+
+  // if deployment type is sandbox, use id in the format: resourceName-YYYY-MM-DD to avoid multiple deployments in the same day
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:261]: Same point here as above: why partition by date?

File: packages/amplify-graphql-transformer-core/API.md [hunk 4]
```diff
@@ -309,7 +310,7 @@ export class GraphQLTransform {
     // Warning: (ae-forgotten-export) The symbol "TransformOption" needs to be exported by the entry point index.d.ts
     //
     // (undocumented)
-    transform({ assetProvider, dataSourceStrategies, nestedStackProvider, parameterProvider, rdsLayerMapping, rdsSnsTopicMapping, schema, scope, sqlDirectiveDataSourceStrategies, synthParameters, }: TransformOption): void;
+    transform({ assetProvider, dataSourceStrategies, nestedStackProvider, parameterProvider, rdsLayerMapping, rdsSnsTopicMapping, schema, scope, sqlDirectiveDataSourceStrategies, synthParameters, deploymentIdentifier, }: TransformOption): void;
```
Comment (palpatim) [inline packages/amplify-graphql-transformer-core/API.md:313]: Should we make `deploymentIdentifier` an attribute of `synthParameters` rather than dangling it off the end of the transform input?

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts [hunk 5]
```diff
@@ -195,6 +197,7 @@ export class GraphQLTransform {
     scope,
     sqlDirectiveDataSourceStrategies,
     synthParameters,
+    deploymentIdentifier,
```
Comment (palpatim) [inline packages/amplify-graphql-transformer-core/src/transformation/transform.ts:200]: as above, should `deploymentIdentifier` be part of `synthParameters` rather than a top-level attribute of `TransformOption`?

File: packages/amplify-graphql-transformer-interfaces/API.md [hunk 6]
```diff
@@ -146,6 +146,16 @@ export interface AssetProvider {
     provide: (scope: Construct, name: string, props: AssetProps) => S3Asset;
 }
 
+// @public (undocumented)
+export interface BranchIdentifier {
```
Comment (palpatim) [inline packages/amplify-graphql-transformer-interfaces/API.md:150]: See note above about naming. While the internal interfaces don't strictly have to correspond to the external interfaces, in cases where the serve the same purpose with the same data, it reduces our confusion to keep the names the same.

File: packages/amplify-graphql-transformer/src/graphql-transformer.ts [hunk 7]
```diff
@@ -118,6 +119,7 @@ export type ExecuteTransformConfig = TransformConfig &
     parameterProvider?: TransformParameterProvider;
     assetProvider: AssetProvider;
     synthParameters: SynthParameters;
+    deploymentIdentifier?: DeploymentIdentifier;
```
Comment (palpatim) [inline packages/amplify-graphql-transformer/src/graphql-transformer.ts:122]: This file only applies to Gen 1 transformer right? Is `deploymentIdentifier` required for Gen 1? (Or I might be misremembering how this file is used!)

File: packages/amplify-graphql-model-transformer/src/resolvers/index.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
 export * from './generators';
 export * from './common';
+export * from './rds';
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/index.ts:3]: Why are we now exporting this? (I'm assuming it's to make new logic in `resolver.ts` easier to access somewhere?

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts (hunk 2 above)
Comment (bobbyu99) [reply packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:216]: I was discussing with Christ, Phani, Dane, and Josef, the major concern we have is that, if we use a fixed name like the resourceName, it will not pick up the weekly release/update that might contain security patches. Between weekly and daily, we decided to go with daily. 

I'm still trying to understand the nuances here. Assuming we have a layer version update at noon, and a user deploys once in the morning, and once in the afternoon. 
1) For the current approach using Date.now(), this user will have the update
2) For "per day" approach, this user will not have the update. Unless they deploy again the next day
3) For fixed name approach with only `${resourceName}`, they will never get this update no matter how many times they deploy (i'm not sure if i get it right here)

File: packages/amplify-graphql-model-transformer/src/resolvers/index.ts (hunk 8 above)
Comment (bobbyu99) [reply packages/amplify-graphql-model-transformer/src/resolvers/index.ts:3]: I exported `createLayerVersionCustomResource`, `createSNSTopicARNCustomResource` for the unit test in the same package:
`import { createLayerVersionCustomResource, createSNSTopicARNCustomResource } from '@aws-amplify/graphql-model-transformer';
`
I guess define a relative path instead of exporting is a better choice here?

Comment (palpatim) [reply packages/amplify-graphql-model-transformer/src/resolvers/index.ts:3]: No need to change, as long as we don't need to export it to the public interface. I was just wondering why the new export :)

File: packages/amplify-graphql-api-construct/src/types.ts [hunk 9]
```diff
@@ -786,8 +786,35 @@ export interface AmplifyGraphqlApiProps {
    * For more information, refer to https://docs.amplify.aws/lib/datastore/getting-started/q/platform/js/
    */
   readonly dataStoreConfiguration?: DataStoreConfiguration;
+
+  /**
+   * The deployment identifier for the current deployment.
+   */
+  readonly deploymentIdentifier?: DeploymentIdentifier;
```
[the remark this answers is not in the document]
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/types.ts:793]: `translationBehavior` only contains boolean props and have 1-on-1 parity with feature flags. I was going with that idea initially, but I feel its not very appropriate. Is there any other place we can put `deploymentIdentifier`?

File: packages/amplify-graphql-transformer/src/graphql-transformer.ts (hunk 7 above)
Comment (bobbyu99) [reply packages/amplify-graphql-transformer/src/graphql-transformer.ts:122]: This file is used in both Gen 1 and Gen 2. But `deploymentIdentifier` is not required for Gen 1

File: packages/amplify-graphql-api-construct/src/types.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/types.ts:793]: `translationBehavior` may not be the right place then. I still object to this being part of the API definition though — it’s not definition, it’s context. That context happens to change the way the resource name is generated, but it’s similar to how CDK uses context to control, e.g., regions, account IDs, or dependent resource IDs.

If we don’t use `translationBehavior`, I’m wondering if we can use a shell environment variable or similar to control the transformation behavior.

If that doesn’t work, then we should create a new optional struct inside the GraphQL API construct to hold these flags — something like `transformationContext` or similar. The use case is that they should only ever be used by the Gen2 CLI. We’d mark the entire struct as experimental, or internal, or something, and put big scary warnings on it that customers shouldn’t rely on the behavior directly.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 10]
```diff
@@ -275,6 +284,27 @@ export class AmplifyGraphqlApi extends Construct {
...
   }
 
+  /**
+   * Gets the deployment information from the context.
+   * @returns the deployment identifier.
+   */
+  private getDeploymentIdentifier(): DeploymentIdentifier {
+    const deploymentType = this.node.tryGetContext(CDKContextKey.DEPLOYMENT_TYPE);
+
+    if (deploymentType === 'sandbox') {
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:294]: Consider abstracting the 'sandbox' magic string into a centralized constant.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 11]
```diff
@@ -275,6 +284,27 @@ export class AmplifyGraphqlApi extends Construct {
...
+      return {
+        deploymentType: deploymentType,
+        namespace: this.node.tryGetContext(CDKContextKey.BACKEND_NAMESPACE),
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:297]: What is the behavior if `BACKEND_NAMESPACE` context key isn't set? Will there be a meaningful error propagated to a customer-facing context?

..
..
..

Trick question :). Even if the error is something like "backendNamespace context key not set", that doesn't tell the customer how to fix it. We should catch the condition ourselves -- if either of the context keys are set, both must be set, and we should craft an appropriate error message to help customers resolve the situation.)

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 12]
```diff
@@ -275,6 +284,27 @@ export class AmplifyGraphqlApi extends Construct {
...
+      } as SandboxDeploymentIdentifier;
+    } else {
+      return {
+        deploymentType: deploymentType,
+        namespace: this.node.tryGetContext(CDKContextKey.BACKEND_NAMESPACE),
+        name: this.node.tryGetContext(CDKContextKey.BACKEND_NAME),
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:303]: What if one or both of the context key isn't set? Maybe it's always set for a Gen 2 CLI invocation, but customers can also do a straight-up `cdk synth` or `cdk deploy` without setting context.

File: packages/amplify-graphql-model-transformer/API.md [hunk 13]
```diff
@@ -5,7 +5,10 @@
 ```ts
 
 import { AppSyncDataSourceType } from '@aws-amplify/graphql-transformer-interfaces';
+import { AwsCustomResource } from 'aws-cdk-lib/custom-resources';
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/API.md:8]: I'm surprised at the number of new exports from this package -- are these all only to enable unit testing of new helpers? If so, we should adjust our package-level `index.ts` to ensure that we only vend truly public exports from the package. If not, what other packages are using the new exports? (I'm assuming these aren't intended for customer consumption?)

File: packages/amplify-graphql-model-transformer/src/__tests__/custom-resources.test.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  const snsTopicARNCustomResourceType = 'Custom::SQLSNSTopicARNCustomResource';
+
+  let app: App;
+  let stack: Stack;
+  beforeEach(() => {
+    app = new App();
+    stack = new Stack(app, 'TestStack');
+  });
+
+  it('should create a layer version custom resource with correct physical ID format for sandbox deployment', () => {
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/__tests__/custom-resources.test.ts:31]: Do we have equivalent tests for branch deployments?

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts [hunk 15]
```diff
@@ -216,7 +230,7 @@ export const createLayerVersionCustomResource = (scope: Construct, resourceNames
       },
       // Make the physical ID change each time we do a deployment, so we always check for the latest version. This means we will never have
       // a strictly no-op deployment, but the SQL Lambda configuration won't change unless the actual layer value changes
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:232]: nit: move this comment above where you're creating they physical ID so we're explaining the motivation in context.

File: packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts [hunk 16]
```diff
@@ -251,7 +279,7 @@ export const createSNSTopicARNCustomResource = (scope: Construct, resourceNames:
...
```
Comment (palpatim) [inline packages/amplify-graphql-model-transformer/src/resolvers/rds/resolver.ts:281]: As above, move this comment to where you're creating the physical ID

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 17]
```diff
@@ -275,6 +294,62 @@ export class AmplifyGraphqlApi extends Construct {
...
+  /**
+   * Gets the deployment information from the context.
+   * @returns the deployment identifier.
+   */
+  private getDeploymentIdentifier(): DeploymentIdentifier {
+    const deploymentType = this.node.tryGetContext(CDKContextKey.DEPLOYMENT_TYPE);
+    const expectedDeploymentTypes = [this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE];
+
+    // Throw an error if the deployment type is neither sandbox nor branch.
+    if (!expectedDeploymentTypes.includes(deploymentType)) {
```
Comment (edwardfoyle) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:306]: If the GQL construct is used outside of the context of an Amplify backend (ie as a construct in a native CDK project), then these context props will not be present

File: packages/amplify-graphql-transformer-interfaces/src/transformer-context/synth-parameters.ts [hunk 18]
```diff
@@ -7,4 +7,18 @@ export type SynthParameters = {
   identityPoolId?: string;
   adminRoles?: string[];
   enableIamAccess?: boolean;
+  deploymentIdentifier?: DeploymentIdentifier;
 };
+
+export interface SandboxDeploymentIdentifier {
```
Comment (edwardfoyle) [inline packages/amplify-graphql-transformer-interfaces/src/transformer-context/synth-parameters.ts:13]: branch and sandbox deployments both have a name and namespace:
https://github.com/aws-amplify/amplify-backend/blob/39175772301cd9db58fd07ff6ed3e2c0e20fb876/packages/plugin-types/API.md?plain=1#L64-L74

In the case of sandbox, the name is the value of `package.json#name` and the namespace is either the local username or the value of `--identifier` supplied to the CLI

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 17 above)
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:306]: To clarify, you are referring to the class variables `this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE` right?

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 19]
```diff
@@ -275,6 +294,62 @@ export class AmplifyGraphqlApi extends Construct {
...
+        `${CDKContextKey.DEPLOYMENT_TYPE} must be one of (${expectedDeploymentTypes.join(', ')}). ` +
+        'Please set the deployment type by running:\n' +
+        'cdk --context amplify-backend-type=sandbox\n' +
+        'or\n' +
+        'cdk --context amplify-backend-type=branch'
+      );
+    }
+
+    if (deploymentType === this.SANDBOX_DEPLOYMENT_TYPE) {
+      const namespace = this.node.tryGetContext(CDKContextKey.BACKEND_NAMESPACE);
```
[the remark this answers is not in the document]
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:317]: Sorry im not quite getting the question here. I'm getting the deployment type, the name and namespace to construct a DeploymentIdentifier to be used later. 

But as you mention in the other comment above that sandbox also has both name and namespace, i dont need a `if ... else ...`. Regardless of sandbox or branch deployments, i gotta make sure that both name and namespace are present.

[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:317]: It doesn't look like you need to pass in a whole DeploymentIdentifier though. The only place it's being read (I think) is in `isSandboxDeployment` which only looks at the deployment type, not the other fields. So maybe you could just pass in `deploymentType` and not care about reading any of the other values

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 17 above)
Comment (edwardfoyle) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:306]: Correct, none of these context values would be present in a native CDK project. The construct should probably treat absence of these values the same as a "branch" deployment

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:317]: I see what you mean. Yeah `isSandboxDeployment` only needs `deploymentType`. As I was discussing with my team earlier, we decided to include name and namespace as we think they could be useful information in the future.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 17 above)
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:306]: Let me rephrase what you are suggesting here. Are you saying that in the case of a native CDK project, where `this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE` are absent, `getDeploymentIdentifier()` should treat this case as a branch deployment and return a `BranchDeploymentIdentifier`?

I see another workaround for the absence of `this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE` is to make them local inside `getDeploymentIdentifier()`. Will this solve the problem?

Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:301]: You should only emit the warning if a deployment type is actually present:

```
if (deploymentType && !expectedDeploymentTypes.includes(deploymentType)) {
```

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 20]
```diff
@@ -302,52 +297,35 @@ export class AmplifyGraphqlApi extends Construct {
...
-        'or\n' +
-        'cdk --context amplify-backend-type=branch'
+      console.warn(
+        `${CDKContextKey.DEPLOYMENT_TYPE} is recommended to be one of (${expectedDeploymentTypes.join(', ')}).\n` +
+          'Consider setting the deployment type by running:\n' +
+          'cdk --context amplify-backend-type=sandbox\n' +
+          'or\n' +
+          'cdk --context amplify-backend-type=branch\n',
       );
     }
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:309]: You should go ahead and return if we encounter this condition, since we can't safely proceed with an unknown deployment type.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 21]
```diff
@@ -275,6 +289,45 @@ export class AmplifyGraphqlApi extends Construct {
...
+      console.warn(
+        `${CDKContextKey.DEPLOYMENT_TYPE} is recommended to be one of (${expectedDeploymentTypes.join(', ')}).\n` +
+          'Consider setting the deployment type by running:\n' +
+          'cdk --context amplify-backend-type=sandbox\n' +
+          'or\n' +
+          'cdk --context amplify-backend-type=branch\n',
+      );
+    }
+
+    // Check namespace and name. Throw an error, if ONLY one of them is set, .
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:311]: punctuation nit :)

```suggestion
    // Check namespace and name. Throw an error if ONLY one of them is set.
```

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 20 above)
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:309]: You should also return if we don't have a deployment type at all, right?

Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:309]: > You should go ahead and return if we encounter this condition, since we can't safely proceed with an unknown deployment type.

If we return here on an unknown deployment type, we would not be checking namespace and name right?

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 17 above)
Comment (sobolk) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:301]: As far as I recall data construct was supposed to be usable outside of amplify in raw CDK apps.

Should this logic be 
```
if (deploymentType) {
  if(deploymentType is invalid) {
   // If we're in Amplify Gen2 app, customer doesn't control these keys, fail fast.
   throw Error();
  }
} else {
   // you are in raw CDK app, none of Amplify CDK context keys is present, which is fine, default to something reasonable.
}
```

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 20 above)
Comment (bobbyu99) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:309]: > You should also return if we don't have a deployment type at all, right?

Yeah i guess in that case returning an empty deploymentIdentifier would be appropriate

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts]: I'm not going to block on it, but my recommendation is YAGNI. It is easy enough to add it in the future that I would punt until necessary.

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 22]
```diff
@@ -275,6 +289,45 @@ export class AmplifyGraphqlApi extends Construct {
...
+    const expectedDeploymentTypes = [this.SANDBOX_DEPLOYMENT_TYPE, this.BRANCH_DEPLOYMENT_TYPE];
+
+    // Log a warning if the deployment type is neither sandbox nor branch.
+    if (!expectedDeploymentTypes.includes(deploymentType)) {
+      console.warn(
+        `${CDKContextKey.DEPLOYMENT_TYPE} is recommended to be one of (${expectedDeploymentTypes.join(', ')}).\n` +
+          'Consider setting the deployment type by running:\n' +
+          'cdk --context amplify-backend-type=sandbox\n' +
+          'or\n' +
+          'cdk --context amplify-backend-type=branch\n',
```
Comment (edwardfoyle) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:307]: I think this warning is going to cause more confusion than clarity. We don't intend for customers to interact directly with these context values so they aren't going to know what this mean. Also, saying "deployment type is recommended to be... consider setting..." is a lot of weasel words and immediately raises the question "what are the implications of setting or not setting?". If I saw this warning I would just be confused and probably ignore it

Comment (edwardfoyle) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:307]: We also already have [this mechanism](https://github.com/aws-amplify/amplify-backend/blob/d8b43d252c7d5553bf719dd889571807098adf13/packages/backend-data/src/factory.ts#L264-L272) for changing GSI update behavior in sandbox vs non-sandbox. Perhaps we could follow a similar pattern to inject necessary context into the data construct rather than the construct trying to fetch it.

**Review by palpatim — DISMISSED** [review]

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts [hunk 23]
```diff
@@ -275,6 +294,89 @@ export class AmplifyGraphqlApi extends Construct {
...
+    if (sandboxDeploymentIdentifier) {
+      return sandboxDeploymentIdentifier;
+    }
+
+    const branchDeploymentIdentifier = this.makeBranchDeploymentIdentifier(deploymentType);
+    if (branchDeploymentIdentifier) {
+      return branchDeploymentIdentifier;
+    }
+
+    console.warn(`Encountering an unknown deployment type: ${deploymentType}.\n`);
```
Comment (palpatim) [inline packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts:317]: nit (for future PR is fine): Change verbiage, remove redundant newline: 
```suggestion
    console.warn(`Encountered an unknown deployment type: ${deploymentType}.`);
```

File: packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts (hunk 22 above)
Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts]: > We also already have [this mechanism](https://github.com/aws-amplify/amplify-backend/blob/d8b43d252c7d5553bf719dd889571807098adf13/packages/backend-data/src/factory.ts#L264-L272) for changing GSI update behavior in sandbox vs non-sandbox. Perhaps we could follow a similar pattern to inject necessary context into the data construct rather than the construct trying to fetch it?

We'd actually prefer to go the opposite direction, and use sandbox detection to drive the GSI update behavior, rather than make CLI flags privy to more of the transformer internals. We'd have to use aspects to hook onto a specific part of the target stack (e.g., the code you link has to know about the existence of `AmplifyDynamoDbTableWrapper`), and so the CLI now has to be tightly coupled to some part of the underlying stack. Context seems to me the right way for specifying environmental information, and to let the consuming stack do whatever it needs to.

Comment (edwardfoyle) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts]: I think the "most correct" thing to do is to expose this behavior and the GSI behavior as a prop on the DataConstruct. Then the `backend-data` package can decide how to instantiate the construct based on the Amplify context it knows about and CDK consumers can do the same. This solution isn't eliminating tight coupling, it's just moving it. You are depending on the name and values of the CDK context set by `amplify-backend` without an explicit interface guarding it.

Comment (palpatim) [reply packages/amplify-graphql-api-construct/src/amplify-graphql-api.ts]: Fair point about "moving the coupling without an interface."
